### PR TITLE
feat: add time series pool

### PR DESCRIPTION
**Pull request by cc004**
please read [rfc](https://github.com/4paradigm/rfcs/pull/6) for more details.

### CODE REVIEW

File: src/base/time_serise_pool.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            pool_.erase(pair);
+        }
+    }
+    bool Empty() { return pool_.empty(); }
+
+ private:
+    // key is the time / (60 * 60 * 1000)
+    uint32_t block_size_;
+    std::map<uint32_t, std::unique_ptr<TimeBucket>> pool_;
+    inline static uint32_t keyof(uint64_t time) { return time / (60 * 60 * 1000); }
```
Comment (dl239) [inline src/base/time_serise_pool.h]: Rename to ComputeTimeSlot

File: src/base/time_serise_pool.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            auto bucket = new TimeBucket(block_size_);
+            pool_.insert(
+                std::pair<uint32_t, std::unique_ptr<TimeBucket>>(keyof(time), std::unique_ptr<TimeBucket>(bucket)));
+            return bucket->Alloc(size);
+        }
+
+        return pair->second->Alloc(size);
+    }
+    void Free(uint64_t time) {
+        auto pair = pool_.find(keyof(time));
```
Comment (dl239) [inline src/base/time_serise_pool.h]: judge iterator's validation before used

File: src/base/time_serise_pool.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+class TimeSerisePool {
+ public:
+    explicit TimeSerisePool(uint32_t block_size) : block_size_(block_size) {}
+    void* Alloc(uint32_t size, uint64_t time) {
+        auto pair = pool_.find(keyof(time));
+        if (pair == pool_.end()) {
+            auto bucket = new TimeBucket(block_size_);
+            pool_.insert(
+                std::pair<uint32_t, std::unique_ptr<TimeBucket>>(keyof(time), std::unique_ptr<TimeBucket>(bucket)));
```
Comment (dl239) [inline src/base/time_serise_pool.h]: need not compute time slot again

File: src/base/time_serise_pool.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+namespace base {
+
+class TimeBucket {
+ public:
+    explicit TimeBucket(uint32_t block_size)
+        : block_size_(block_size), current_offset_(block_size + 1), object_num_(0) {
+        head_ = reinterpret_cast<Block*>(malloc(sizeof(Block*)));  // empty block at end
+        head_->next = NULL;
+    }
+    void Clear() {
```
Comment (imotai) [inline src/base/time_serise_pool.h]: 缺少析构函数

File: src/base/time_serise_pool.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        return pair->second->Alloc(size);
+    }
+    void Free(uint64_t time) {
+        auto pair = pool_.find(ComputeTimeSlot(time));
+        if (pair != pool_.end() && pair->second->Free()) {
+            pool_.erase(pair);
+        }
+    }
+    bool Empty() { return pool_.empty(); }
+
```
Comment (imotai) [inline src/base/time_serise_pool.h]: 加上析构函数，直接删除所有的bucket

Comment (cc004) [reply src/base/time_serise_pool.h]: 这里是不是不需要加析构函数，因为在map析构的时候unique_ptr也会被析构，会自动调用TimeBucket的delete和析构

**Comment by vagetablechicken** [comment]
Time serise? Is it a spelling mistake? It should be time series.

File: src/base/skiplist.h [hunk 6]
```diff
@@ -269,6 +296,7 @@ class Skiplist {
     }
 
     // Need external synchronized
+    // called iif skiplist is using tcalloc
```
Comment (dl239) [inline src/base/skiplist.h]: `if` and `tcmalloc` spell error

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Pull request closed without merging**